File: cpp/03-operators/operators_logical.cpp
#include <iostream>

int main() {
    int age = 25;
    int score = 88;

    std::cout << "age == 25 : " << (age == 25) << '\n';
    std::cout << "age != 30 : " << (age != 30) << '\n';
    std::cout << "score >= 90: " << (score >= 90) << '\n';
    std::cout << "score < 90 : " << (score < 90) << '\n';

    bool eligible = (age >= 18) && (score >= 80);
    bool flagged  = (age < 18) || (score < 60);

    std::cout << "eligible (age>=18 && score>=80) = " << eligible << '\n';
    std::cout << "flagged  (age<18  || score<60)  = " << flagged  << '\n';
    std::cout << "!eligible = " << !eligible << '\n';

    // Short-circuit: the right side is never evaluated when the left
    // already determines the result, so divide-by-zero is avoided.
    int divisor = 0;
    if (divisor != 0 && (100 / divisor) > 1) {
        std::cout << "branch taken\n";
    } else {
        std::cout << "short-circuit prevented division by zero\n";
    }

    return 0;
}

File: matlab/03-operators/operators_logical.m
% Logical operators
a = true;
b = false;

% Short-circuit (scalars only): && and ||
fprintf('a && b = %d\n', a && b);
fprintf('a || b = %d\n', a || b);
fprintf('~a     = %d\n', ~a);     % Logical NOT

% Element-wise logical on arrays
v1 = [1 0 1 0];
v2 = [1 1 0 0];
disp('v1 & v2:'); disp(v1 & v2)
disp('v1 | v2:'); disp(v1 | v2)
disp('xor(v1,v2):'); disp(xor(v1, v2))

% Short-circuit prevents evaluating the right side
x = 0;
if x ~= 0 && (10 / x) > 1
    disp('safe division')
else
    disp('skipped division by zero')
end
